File: macclean/cleaners/privacy.py
import sqlite3
from pathlib import Path
import click
from rich.console import Console
from rich.table import Table
from rich.panel import Panel

console = Console()

_TCC_SERVICES = {
    "kTCCServiceCamera": "Camera",
    "kTCCServiceMicrophone": "Microphone",
    "kTCCServiceScreenCapture": "Screen Recording",
    "kTCCServiceLocation": "Location",
    "kTCCServiceAddressBook": "Contacts",
    "kTCCServiceCalendar": "Calendar",
    "kTCCServicePhotos": "Photos",
    "kTCCServiceSystemPolicyAllFiles": "Full Disk Access",
    "kTCCServiceAccessibility": "Accessibility",
    "kTCCServiceReminders": "Reminders",
    "kTCCServiceLiverpool": "Location Services",
    "kTCCServiceUbiquity": "iCloud",
    "kTCCServiceShareKit": "Sharing",
}

_AUTH = {0: "[red]Denied[/]", 1: "[dim]Unknown[/]", 2: "[green]Allowed[/]", 3: "[yellow]Limited[/]"}


def _query_tcc(db_path: Path) -> list[tuple[str, str, str]]:
    rows = []
    if not db_path.exists():
        return rows
    try:
        conn = sqlite3.connect(f"file:{db_path}?mode=ro", uri=True)
        cur = conn.execute("SELECT service, client, auth_value FROM access ORDER BY service, client")
        for service, client, auth_val in cur.fetchall():
            svc = _TCC_SERVICES.get(service, service.replace("kTCCService", ""))
            auth = _AUTH.get(auth_val, str(auth_val))
            rows.append((svc, client, auth))
        conn.close()
    except Exception as e:
        rows.append(("Error", str(e), ""))
    return rows


@click.command()
@click.pass_context
def cmd(ctx):
    """Show app privacy permissions (camera, mic, screen recording, location, etc.)."""
    home = Path.home()

    for label, db_path in [
        ("User Permissions", home / "Library" / "Application Support" / "com.apple.TCC" / "TCC.db"),
        ("System Permissions", Path("/Library/Application Support/com.apple.TCC/TCC.db")),
    ]:
        rows = _query_tcc(db_path)
        if not rows:
            console.print(f"[dim]{label}: not readable (run with sudo for system DB)[/]")
            continue

        table = Table(show_header=True, show_lines=True)
        table.add_column("Permission")
        table.add_column("App / Bundle ID")
        table.add_column("Status")
        for svc, client, auth in rows:
            table.add_row(svc, client, auth)

        console.print(Panel(table, title=f"[bold cyan]{label}[/]"))

File: src/cleaners/privacy.rs
use anyhow::Result;
use colored::Colorize;
use comfy_table::{Table, presets::UTF8_BORDERS_ONLY};
use crate::core::cmd::run_cmd;
use crate::cleaners::security::map_service;

fn map_auth(v: &str) -> &'static str {
    match v.trim() {
        "0" => "Denied",
        "1" => "Unknown",
        "2" => "Allowed",
        "3" => "Limited",
        _   => "Unknown",
    }
}

fn query_tcc(label: &str, db_path: &std::path::Path) {
    println!("\n{}", format!("[ App Permissions — {} ]", label).cyan().bold());

    if !db_path.exists() {
        println!("  Database not accessible: {}", db_path.display());
        return;
    }

    let db_str = db_path.to_string_lossy();
    let out = run_cmd(&[
        "sqlite3",
        "-separator",
        "|",
        &db_str,
        "SELECT service,client,auth_value FROM access ORDER BY service,client",
    ]);

    if !out.success() || out.output.trim().is_empty() {
        println!("  No entries found (or database not readable).");
        return;
    }

    let mut table = Table::new();
    table.load_preset(UTF8_BORDERS_ONLY);
    table.set_header(vec!["Permission", "App", "Status"]);

    for line in out.output.lines() {
        let parts: Vec<&str> = line.splitn(3, '|').collect();
        if parts.len() < 3 {
            continue;
        }
        let service = map_service(parts[0].trim());
        let client  = parts[1].trim();
        let status  = map_auth(parts[2].trim());

        let status_colored = match status {
            "Allowed" => status.green().to_string(),
            "Denied"  => status.red().to_string(),
            _         => status.yellow().to_string(),
        };

        table.add_row(vec![
            service.to_string(),
            client.to_string(),
            status_colored,
        ]);
    }
    println!("{}", table);
}

pub fn run() -> Result<()> {
    let home = dirs::home_dir().unwrap_or_else(|| std::path::PathBuf::from("."));

    let user_db = home.join("Library/Application Support/com.apple.TCC/TCC.db");
    let sys_db  = std::path::PathBuf::from("/Library/Application Support/com.apple.TCC/TCC.db");

    query_tcc("User", &user_db);
    query_tcc("System", &sys_db);

    Ok(())
}
